### PR TITLE
crimson/osd: Osd UUID zero

### PR DESCRIPTION
**Pull request by liu-chunmei**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by athanatos** [comment]
What is supposed to call FuturizedStore::set_fsid()?

File: src/crimson/os/seastore/seastore.cc [hunk 1]
```diff
@@ -188,17 +189,22 @@ seastar::future<> SeaStore::umount()
...
   LOG_PREFIX(SeaStore::write_fsid);
-  return read_meta("fsid").then([this, FNAME, new_osd_fsid] (auto tuple) {
+  return read_meta("fsid").then([this, FNAME, new_osd_fsid] (auto tuple) mutable {
     auto [ret, fsid] = tuple;
     std::string str_fsid = stringify(new_osd_fsid);
     if (ret == -1) {
+       if (new_osd_fsid.is_zero()) {
+         new_osd_fsid.generate_random();
+       }
+       set_fsid(new_osd_fsid);
```
Comment (athanatos) [inline src/crimson/os/seastore/seastore.cc:199]: Instead of doing this here, can we just randomize it at the beginning of mkfs?

**Comment by liu-chunmei** [comment]
> What is supposed to call FuturizedStore::set_fsid()?

it is objectstore interface, which is called in osd::mkfs,  but we didn't implement like classic osd, so crimson-osd not call it in osd::mkfs.

File: src/crimson/os/seastore/seastore.cc (hunk 1 above)
Comment (liu-chunmei) [reply src/crimson/os/seastore/seastore.cc:199]: yes, we can, will do it then set_fsid will be called in osd::mkfs.

**Comment by athanatos** [comment]
OSD::mkfs calls set_fsid and then immediately calls mkfs.  For crimson/FuturizedStore, we are instead passing the fsid as an argument to mkfs.  We shouldn't need a set_fsid interface method at all, I don't think.

File: src/crimson/os/seastore/seastore.cc [hunk 2]
```diff
@@ -205,11 +206,14 @@ seastar::future<> SeaStore::write_fsid(uuid_d new_osd_fsid)
 
 SeaStore::mkfs_ertr::future<> SeaStore::mkfs(uuid_d new_osd_fsid)
 {
-  return read_meta("mkfs_done").then([this, new_osd_fsid] (auto tuple) {
+  return read_meta("mkfs_done").then([this, new_osd_fsid] (auto tuple) mutable {
```
Comment (tchaikov) [inline src/crimson/os/seastore/seastore.cc:209]: i'd suggest fix the caller instead. in the case of teuthology test, i assume that the only possible caller is `main()` in `crimson/osd/main.cc`. so `osd_uuid` should have been specified at command line, right?

could you paste the pulpito link or point me to the test suite ? so i can take a look?

Comment (athanatos) [reply src/crimson/os/seastore/seastore.cc:209]: I think the solution is really just to randomize a 0 fsid in SeaStore::mkfs.

Comment (liu-chunmei) [reply src/crimson/os/seastore/seastore.cc:209]: > i'd suggest fix the caller instead. in the case of teuthology test, i assume that the only possible caller is `main()` in `crimson/osd/main.cc`. so `osd_uuid` should have been specified at command line, right?
> 
> could you paste the pulpito link or point me to the test suite ? so i can take a look?

it is intel inner teuthology test environment. 
uuid is not set in command line when doing mkfs.   
 sudo MALLOC_CHECK_=3 adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage ceph-osd --no-mon-config --cluster ceph --mkfs --mkkey -i 0 --monmap /home/ubuntu/cephtest/ceph.monmap

Comment (tchaikov) [reply src/crimson/os/seastore/seastore.cc:209]: i see. can we do this in `main.cc` instead? IMHO, SeaStore is supposed to do whatever it is told to do. changing the input parameter, or the captured parameter in this case, could be confusing. like
```c++
if (config.count("mkfs")) {
  auto osd_uuid = local_conf().get_val<uuid_d>("osd_uuid");
  if (osd_uuid.is_zero()) {
    // use a random osd uuid if not specified
    osd_uuid.generate_random();
  }
  osd.invoke_on(...);
}
```

File: src/crimson/os/seastore/seastore.cc [hunk 3]
```diff
@@ -205,11 +206,14 @@ seastar::future<> SeaStore::write_fsid(uuid_d new_osd_fsid)
...
     auto [done, value] = tuple;
     if (done == 0) {
       return seastar::now();
     } else {
+      if (new_osd_fsid.is_zero()) {
+        new_osd_fsid.generate_random();
```
Comment (rzarzynski) [inline src/crimson/os/seastore/seastore.cc:215]: Hmm, in the classical OSD this logic somehow became store-specific. In `BlueStore`:

```cpp
  void set_fsid(uuid_d u) override {
    fsid = u;
  }
  uuid_d get_fsid() override {
    return fsid;
  }
```

```cpp
int BlueStore::mkfs()
{
  // ...
  r = _read_fsid(&old_fsid);
  if (r < 0 || old_fsid.is_zero()) {
    if (fsid.is_zero()) {
      fsid.generate_random();
      dout(1) << __func__ << " generated fsid " << fsid << dendl;
    } else {
      dout(1) << __func__ << " using provided fsid " << fsid << dendl;
    }
    // we'll write it later.
  } else {
```

Comment (liu-chunmei) [reply src/crimson/os/seastore/seastore.cc:215]: yes, it is test uuid zero in store level.

**Comment by liu-chunmei** [comment]
jenkins please retests ceph API tests.

File: src/crimson/os/seastore/seastore.cc (hunk 2 above)
Comment (athanatos) [reply src/crimson/os/seastore/seastore.cc:209]: Yeah, I agree with this change.  There's no reason for this to be store specific -- it's really a property of the behavior of the osd_uuid config.  Let's change this in main.cc as suggested here.

**Review by tchaikov — APPROVED** [review]
might want to improve the title of the commit message to summarize the change, something like

> crimson/osd: randomize the osd_uuid if not specified
>
> this mirrors the behavior of classic osd and addresses the failure 
> spotted in teuthology based test:
> sudo ceph --cluster ceph osd new 00000000-0000-0000-0000-000000000000 0
>

**Review by athanatos — APPROVED** [review]

**Pull request merged**